### DISCUSSION TITLE
Could anybody tell me what the LoadBlocker's working principle is ?

### DISCUSSION THREAD

**Discussion by boluochoufeng** [servo/servo#34590] [Q&A] [answered]
I am a Rust beginner and have recently been learning about the Servo project. I came across the `LoadBlocker` structure in `document_loader.rs`, but I don't understand why it is needed and how it works. Could someone please help me with this? 
Thank you!

**Comment by jdm** [comment, marked as answer]
LoadBlocker is an implementation of the concept of ["delaying the load event"](https://html.spec.whatwg.org/multipage/parsing.html#delay-the-load-event) from the HTML specification. As long as any LoadBlockers are registered for a document, the document's load event will be prevented. When a load completes, the element that initiated it has the opportunity to remove a load blocker that corresponds to the load.

**Reply by boluochoufeng** [reply]
Thanks for your answer! 
These are my understandings: 
1. The `LoadBlocker` can be used to add `LoadType` to the `DocumentLoader`. 
2. When `finish_load` is called, it checks whether the `DocumentLoader` is still `blocked (is_blocked())`. If it is, it returns. Otherwise, it marks the current document as unblocked. In the next event handling loop of the script_thread, this document has a chance to start rendering. 

Is there anything wrong with my understanding? Additionally, I noticed that elements like `img`, `stylesheet`, and `video` also manipulate the `LoadBlocker`. If my understanding is correct, then the rendering of the document will be blocked by their loading, which could lead to a poor user experience?

**Reply by jdm** [reply]
Your understanding is correct. And yes, those elements all block the "load" event being fired at the document, as the HTML specification requires. I don't think it's correct to describe it as a poor user experience.

**Reply by jdm** [reply]
The rendering of the document is independent of the load event.

**Reply by boluochoufeng** [reply]
OK. Thanks for your answer again. It seems I need to study the specification.
